Fix #1468: Retain Drag and Drop state after an incorrect answer

Rd4dev opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Fix #1468 

### This PR includes

- Utilized `explorationProgressController` to retrieve the **'wrongAnswerList'** from ephemeralState's pendingState and use it to populate the `DragDropInteractionContentViewModel` list.
- When `wrongAnswerList` is available, corresponding content IDs and HTML values are applied. If not, the original `contentIdHtmlMap` and `SubtitledHtml` values are used as a fallback.
- The `choiceItems` are set up as LiveData to ensure the UI can react to changes.
- The `_choiceItems` are copied to `_originalChoiceItems ` for detecting arrangement box errors during submission when the answer has not changed.
- Retains linked-merged / unlinked states based on the `wrongAnswerList` data

## For UI-specific PRs only

### Re-order Retain State

https://github.com/user-attachments/assets/f4c3fd72-d96d-4ee8-b82e-af134ea06a2b

### Group / Unlink Retain

https://github.com/user-attachments/assets/c90a5fc2-ccf9-4e21-a7cd-31f9ec1bd6c0

### Espresso Tests

![image](https://github.com/user-attachments/assets/2e3617eb-4d28-4bee-b8a9-f94c2a3b0231)

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

Rd4dev committed:
Fix #1468: Retain Drag and Drop state after an incorrect answer

Rd4dev committed:
Removed unused import

Yet to figure out if runBlocking is the suitable approach and then continue with figuring out the tests

Rd4dev committed:
Fix Ktlint issues

Rd4dev commented:
@BenHenning, @adhiamboperes, could you PTAL and confirm if this could effectively retain the drag-and-drop states? 
I'm uncertain if using `runBlocking` is the right approach here. 

```
runBlocking {
  when (val result = explorationProgressController.getCurrentState().retrieveData()) { ...}
}
```

#

I'll rework this (possibly by finding a way to update the adapter) since it's blocking the current thread and causing test timeouts.

Rd4dev committed:
Updating the UI via live data - reset works; the empty input error recognition fails

The implementation changes the initial value, so the original value is not equal to the or will never be equal to the choice items (other than the actual original reordering) since it is continuing from the retain state. So either this is the way to go or should figure out a way to set original item as retained value (which I doubt if could be properly synced as it is asynchronous and the first return is a empty list).

Rd4dev committed:
Updated the original choice items with a new instance of the choice Items

This works well to check for the get Submit Time Error checks with the updated ephemeral state value as needed now the only step I is to make sure the _choiceItems is handled properly and also doubt if originalItems would even need any separate functionality.

Rd4dev committed:
Initializing the _choiceItems to a empty list to never have a null reference

The original choice item list would be still required as it is required as a fallback value for the _choiceItems. Running the tests resulted in 2 failures and they need to be addressed.

Rd4dev committed:
Debugging save lobby before swiching to test in base branch

Rd4dev committed:
Temporary Fix to handle grouping - resets (does not retain)

A way to handle the grouping of elements with the wrongAnswerList is yet to be handled. Since choiceString points to the original size and list it does not align with the wrongAnswerList and also the wrongAnswerList and updateList would need a different handling - (to be figured out)

Rd4dev committed:
Implementation of a retaining the reordered state with linking / grouping

Needs a refactoring to combine them as one conditional return list and also to check the updation of the list on each drag end (doesn't occur with unlinkable elements).

Rd4dev committed:
Refactored the result retrieval with wrongAnswerList and added Robolectric test cases for merge and unlink drag drop elements

Rd4dev committed:
Ktlint indentation fix

Rd4dev committed:
Added test case to check the retained wrong Answer list text when merged / linked

Rd4dev commented:
@adhiamboperes, utilized LiveData to retain the states and added corresponding Robolectric tests. However, I haven't been able to add or verify the Espresso tests with issues, as mentioned in #5435. Is there anything I missed or any workaround to resolve this? Could you PTAL.

Vishwajith-Shettigar commented:
> @adhiamboperes, utilized LiveData to retain the states and added corresponding Robolectric tests. However, I haven't been able to add or verify the Espresso tests with issues, as mentioned in #5435. Is there anything I missed or any workaround to resolve this? Could you PTAL.

The workaround would be to comment out the Oppia test rule in that test class.

Rd4dev committed:
Added/Verified Espresso Tests - bumped Mockitor to 3.11.1

Rd4dev commented:
> The workaround would be to comment out the Oppia test rule in that test class.

@Vishwajith-Shettigar -- thanks for helping out, I commented the Oppia Test Rule, but it led to the `MockitoException` shared in the additional context #5435. After bumping Mockito to version 3.11.1, the tests started working with Espresso!

_Yet to approve the version bump._

Vishwajith-Shettigar commented:
> > The workaround would be to comment out the Oppia test rule in that test class.
> 
> @Vishwajith-Shettigar -- thanks for helping out, I commented the Oppia Test Rule, but it led to the `MockitoException` shared in the additional context #5435. After bumping Mockito to version 3.11.1, the tests started working with Espresso!
> 
> _Yet to approve the version bump._

I usually bump to version 2.27.0 as a workaround, but version 2.22.0 >= may also work well with Java 11.

Rd4dev committed:
Revert mockito version introduced to workaround locally

Rd4dev commented:
> I usually bump to version 2.27.0 as a workaround, but version 2.22.0 >= may also work well with Java 11.

Thanks, @Vishwajith-Shettigar, would try that. It was also mentioned during the meeting that PR #5467 would be revived sometime soon and that would resolve this more thoroughly.

Rd4dev committed:
Using the original choice items as a fall back value if the async result fails to retrieve data, making sure the original data is displayed instead of a empty list

adhiamboperes commented on app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/DragAndDropSortInteractionViewModel.kt:
```diff
@@ -49,11 +56,14 @@ class DragAndDropSortInteractionViewModel private constructor(
   private val translationController: TranslationController,
-  userAnswerState: UserAnswerState
+  userAnswerState: UserAnswerState,
+  private val fragment: Fragment,
```
We typically dont pass fragments or contexts to the viewmodel as this could lead to memory leaks.

Rd4dev replied:
Got it, updated it with `Transformations` as suggested.

adhiamboperes commented on app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/DragAndDropSortInteractionViewModel.kt:
```diff
@@ -324,25 +341,70 @@ class DragAndDropSortInteractionViewModel private constructor(
+
+    return _choiceItems.takeIf { it.isNotEmpty() }
+      ?: _originalChoiceItems.toMutableList()
+  }
```
Instead of observing livedata, you can use `Transformations`. this eliminates the need for the fragment instance.
```suggestion
    val ephemeralStateLiveData: LiveData<MutableList<DragDropInteractionContentViewModel>> by lazy {
      Transformations.map(explorationProgressController.getCurrentState().toLiveData()){result->
        return@map when(result){
          is AsyncResult.Failure -> _originalChoiceItems.toMutableList()
          is AsyncResult.Pending -> _originalChoiceItems.toMutableList()
          is AsyncResult.Success -> {
            val state = result.value
            val wrongAnswerList = state.pendingState.wrongAnswerList

            if (wrongAnswerList.isNotEmpty()) {
              val latestWrongAnswerContentIdList = wrongAnswerList.last()
                .userAnswer
                .answer
                .listOfSetsOfTranslatableHtmlContentIds
                .contentIdListsList

              latestWrongAnswerContentIdList.mapIndexed { index, setOfTranslatableHtmlContentIds ->
                DragDropInteractionContentViewModel(
                  contentIdHtmlMap = contentIdHtmlMap,
                  htmlContent = SetOfTranslatableHtmlContentIds.newBuilder().apply {
                    for (contentIds in setOfTranslatableHtmlContentIds.contentIdsList) {
                      addContentIds(
                        TranslatableHtmlContentId.newBuilder().apply {
                          contentId = contentIds.contentId
                        }
                      )
                    }
                  }.build(),
                  itemIndex = index,
                  listSize = latestWrongAnswerContentIdList.size,
                  dragAndDropSortInteractionViewModel = dragAndDropSortInteractionViewModel,
                  resourceHandler = resourceHandler
                )
              }.toMutableList()
            } else {
              _originalChoiceItems.toMutableList()
            }
          }
        }
      }
    }
```

Rd4dev replied:
Thanks for the suggestion, updated it accordingly.

adhiamboperes reviewed (changes requested):
Thank you @Rd4dev! I have left some suggestions, PTAL.

Rd4dev committed:
Merge branch 'develop' of https://github.com/Rd4dev/oppia-android into drag_drop_retain_state_incorrect_answer

Rd4dev committed:
Using Transformations map to transform the live data

Though this works with live data for _choiceItems and rendering them with the latest wrong answer list, it fails to update the original choice items list and also throws an inital null value (and had a hot fix to it). This was not unit tested. May be trying out switch map could be an option with the asyncrhonous data.

Rd4dev committed:
Utilized Transformations switch map to proide the live data

To remove debugging logs and test other functionalities

Rd4dev committed:
Utilized MediatorLiveData to set a initial Pending value to avoid NPE

Rd4dev committed:
Cleanup of debugging logs and intermediate values

Tests passed on Espresso

Rd4dev committed:
Merge branch 'develop' of https://github.com/Rd4dev/oppia-android into drag_drop_retain_state_incorrect_answer

Rd4dev committed:
Fix ktlint indentation issues

Rd4dev commented:
@adhiamboperes, thanks for the review and the helpful direction. 
I've applied the changes, incorporating some adjustments -- PTAL.

Rd4dev committed:
Fix spelling of variable - answerErrorCategory

adhiamboperes reviewed (approved):
Thanks @Rd4dev! This fix LGTM.

Just a note, however, that the video demos seem to be broken. Could you please re-upload them?

Rd4dev committed:
Merge branch 'develop' of https://github.com/Rd4dev/oppia-android into drag_drop_retain_state_incorrect_answer

Rd4dev commented:
Thanks @adhiamboperes for the review. I've updated the video demos in the description. Could you please check if they're functioning properly now?

adhiamboperes commented:
Thanks @Rd4dev, the recordings play now.

adhiamboperes committed:
Merge branch 'develop' into drag_drop_retain_state_incorrect_answer
